### PR TITLE
LPS-121903 Migrate v2 usages in roles/roles-admin-web

### PR DESCRIPTION
**Pull request by julien**
The <clay:management-toolbar-v2 /> tag has been replaced with the
<clay:management-toolbar /> tag (which uses clay v3).

Creating and deleting roles should work as expected, as well as adding or
removing assignees from these roles.

**Test plan (1)**:

- From the global menu, select the **Control Panel** tab
- Navigate to **Roles**
- Make sure you can:
    - Create and delete Regular Roles
    - Create and delete Site Roles
    - Create and delete Organization Roles
    - Create and delete Asset Library Roles
    - Create and delete Account Roles

**Test Plan (2)**:

- Create a Regular Role
- Select (click) the Regular Role you created previously
- Navigate to the **Assignees** tab
- Make sure you can add assignees from this role
- Make sure you can remove assignees from this role

**Additional notes**
- Submitting this as a draft before forwarding to @drewbrokke so that @ambrinchaudhary , @john-co and @markocikos can review

- This module had an `<aui:script>` that was defining one of the management toolbar's event handler, and at the same time calling that handler depending on values present in session storage. Since this might be hard to understand, [this](https://github.com/liferay/liferay-portal/blob/8b8f62ff92709b96a4e96c2fd4b0efc97785e77c/modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp#L205-L221) is what I'm referring to.
In this change, what I've done is replaced that by using a  "propsTransformer" for the management toolbar, another JS module that defines the `addAssignees` function and a `<liferay-frontend:component>` to check session storage and conditionally call the `addAssignees` function.

I think this is OK and I've actually mentioned it to @drewbrokke to make sure, but I'm adding this here so that everyone following this pull request thread knows why this change was made.

### CODE REVIEW

**Comment by julien** [comment]
Oops

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:relevant

**Comment by wincent** [comment]
> [EDIT]: before I ask for a review, I might re-check the `EditAssignees` component - don't bother looking at this right now

Eagerly awaiting the green light...

still-waiting

**Comment by julien** [comment]
@wincent :joy: working on it: still a few missing things here.

Now it's ready for a review before we forward

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:relevant

File: modules/apps/roles/roles-admin-web/src/main/java/com/liferay/roles/admin/web/internal/display/context/EditRoleAssignmentsManagementToolbarDisplayContext.java [hunk 1]
```diff
@@ -115,10 +114,7 @@ public EditRoleAssignmentsManagementToolbarDisplayContext(
 			() -> {
 				DropdownItem dropdownItem = new DropdownItem();
 
-				dropdownItem.setHref(
-					StringBundler.concat(
-						"javascript:", _renderResponse.getNamespace(),
-						"unsetRoleAssignments();"));
+				dropdownItem.putData("action", "unsetRoleAssignments");
```
Comment (markocikos) [inline modules/apps/roles/roles-admin-web/src/main/java/com/liferay/roles/admin/web/internal/display/context/EditRoleAssignmentsManagementToolbarDisplayContext.java:117]: Nice!

File: modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp [hunk 2]
```diff
@@ -56,12 +56,44 @@ renderResponse.setTitle(role.getTitle(locale));
...
+	"selectAssigneesURL", selectAssigneesURL.toString()
+).build();
+%>
+
+<clay:management-toolbar
 	actionDropdownItems="<%= editRoleAssignmentsManagementToolbarDisplayContext.getActionDropdownItems() %>"
+	additionalProps="<%= additionalProps %>"
 	clearResultsURL="<%= editRoleAssignmentsManagementToolbarDisplayContext.getClearResultsURL() %>"
-	componentId="editRoleAssignmentsManagementToolbar"
+	creationMenu="<%= editRoleAssignmentsManagementToolbarDisplayContext.getCreationMenu() %>"
```
Comment (markocikos) [inline modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp:93]: It's a bit weird that we are adding `creationMenu` here. How did this work before? Maybe it was intentional that there is no menu and we fallback to `onClick` on a button.

File: modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/js/EditRoleAssignmentsManagementToolbarPropsTransformer.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+			url: editRoleAssignmentsURL,
+		});
+	};
+
+	return {
+		...otherProps,
+		onActionButtonClick(event, {item}) {
+			const action = item.data?.action;
+
+			if (action === 'unsetRoleAssignments') {
```
Comment (markocikos) [inline modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/js/EditRoleAssignmentsManagementToolbarPropsTransformer.js:67]: Minor SF nit, no need for `action` variable. The same in few other places in this PR.

File: modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp [hunk 4]
```diff
@@ -118,105 +150,7 @@ renderResponse.setTitle(role.getTitle(locale));
...
-		);
-
-		addAssignees();
-	}
-</aui:script>
\ No newline at end of file
+<liferay-frontend:component
+	context="<%= additionalProps %>"
+	module="js/EditRoleAssignments"
+/>
```
Comment (markocikos) [inline modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp:156]: I'm very curious what other reviewers will say about this!

We discussed at length what's the best way to access a function from both JSP and a `PropsTransformer`. This way, both import it from the same file.

**Review by markocikos — COMMENTED** [review]
@julien Looks good overall, only that `creationMenu` is a bit sus.

File: modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp (hunk 2 above)
Comment (julien) [reply modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp:93]: Hey @markocikos good catch ... I think it's a "left over" something I added at one point that isn't needed.
Let me check and confirm that - and I'll update this PR

File: modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/js/EditRoleAssignmentsManagementToolbarPropsTransformer.js (hunk 3 above)
Comment (julien) [reply modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/js/EditRoleAssignmentsManagementToolbarPropsTransformer.js:67]: Yes true

File: modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp (hunk 4 above)
Comment (jbalsas) [reply modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp:156]: I think this is a creative way to get access from both worlds using the existing infrastructure, kudos!

That being said I'm not super happy with the usage of `liferay-frontend:component`, because [it will output](https://github.com/liferay-frontend/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib/src/main/java/com/liferay/frontend/taglib/servlet/taglib/ComponentTag.java#L196) something like: 

```
Liferay.component(
    'xvfx', 
    new jsEditRoleAssignments.default(...)
    {
        destroyOnNavigate: true,
        portletId: '...'
    }
);
```

So, basically we're just triggering a side-effect here via `new myFunction(data)` and then keeping that "instance" in our register of components.

I definitely wouldn't want this to become a pattern, although obviously we don't have a better one at this point. @izaera worked on the "launcher" concept which I think might be better suited for this since we simply want to run a JS snippet but we don't have an easy way to interface with it at the moment.

Comment (drewbrokke) [reply modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp:156]: If a full component creation is undesirable, I'd be ok with an inline script like what we did in `users-admin-web`: https://github.com/liferay/liferay-portal/blob/876bb19b1412a2c79089737c0f1b827dda6d1554/modules/apps/users-admin/users-admin-web/src/main/resources/META-INF/resources/view.jsp#L300-L311.

Comment (jbalsas) [reply modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp:156]: Yeah, I think a simple script might be better in this case because of the semantics. It's still a bit weird because you have to do something like:

```html
<aui:script require='<%= npmResolvedPackageName +  "js/EditRoleAssignments" %>'>
jsEditRoleAssignments.default({
    ...
});
</aui:script>
```

Comment (drewbrokke) [reply modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp:156]: I actually think in this case we can just inline the whole EditRoleAssignments component. Then we can just import and call the `addAssignees` directly.

Comment (jbalsas) [reply modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp:156]: @drewbrokke, would be worth noticing that `window['<portlet:namespace />openSelectUsersDialog']`... is something we're trying to avoid with some of these changes 🤞

Comment (drewbrokke) [reply modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp:156]: @jbalsas agreed, that pattern would not apply to this case.

**Comment by john-co** [comment]
Hi @julien ,

I was investigating some semantic versioning issues in general today and happened to take a closer look in this PR. I found that the modules-semantic-versioning-jdk8/0 issue may be related to this PR since I can't find that error in upstream. It's odd that it is buried as a common failure. Could you take a look?

```
[exec]   PACKAGE_NAME                                       DELTA      CUR_VER    BASE_VER   REC_VER    WARNINGS  
     [exec] = ================================================== ========== ========== ========== ========== ==========
     [exec]   com.liferay.frontend.taglib.clay.data.set          UNCHANGED  1.1.0      1.1.1      1.1.1      VERSION INCREASE SUGGESTED
     [exec] 	+   version    1.1.0
     [exec] 	-   version    1.1.1
     [exec]   com.liferay.frontend.taglib.clay.data.set.provider UNCHANGED  1.0.0      1.0.1      1.0.1      VERSION INCREASE SUGGESTED
     [exec] 	+   version    1.0.0
     [exec] 	-   version    1.0.1
     [exec] * com.liferay.frontend.taglib.clay.servlet.taglib.display.context MINOR      1.5.0      1.5.0      1.6.0      VERSION INCREASE REQUIRED
     [exec] 	<   interface  com.liferay.frontend.taglib.clay.servlet.taglib.display.context.ManagementToolbarDisplayContext
     [exec] 		+   method     isSearchInputAutoFocus()
     [exec] 			+   return     java.lang.Boolean
     [exec] [Baseline Report] Mode: diff (persisted)
```

**Comment by julien** [comment]
Hey @john-co thanks a lot for looking into this.

We changed the `ManagementToolbarDisplayContext` in [https://github.com/liferay-frontend/liferay-portal/pull/833](https://github.com/liferay-frontend/liferay-portal/pull/833) so I'm guessing it might come from [there](https://github.com/liferay-frontend/liferay-portal/pull/833/files#diff-d69e9e6cdd5459041f79572ad29a83bd332a8a6290549ea8618a08e3a61b4e87R141-R144), the addition of this method looks very related to the error.

File: modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp (hunk 4 above)
Comment (julien) [reply modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp:156]: Hey @jbalsas, @drewbrokke thanks for the feedback.

Initially I wanted to do what is done [here](https://github.com/liferay-frontend/liferay-portal/blob/8b8e03937395d44aba2c99114824e685a6269794/modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/js/EditRoleAssignments.js#L25-L37) inside the management toolbar via a callback, but it seems that we didn't want to add the `onComponentMount` callback back in [https://github.com/liferay-frontend/liferay-portal/pull/826](https://github.com/liferay-frontend/liferay-portal/pull/826) so this is what I came up with .

I knew there was [something](https://github.com/liferay-frontend/liferay-portal/pull/835#discussion_r583019982) I hadn't taken into account.

Comment (jbalsas) [reply modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp:156]: > Initially I wanted to do what is done here inside the management toolbar via a callback, but it seems that we didn't want to add the onComponentMount callback back in #826 so this is what I came up with .

I think that wouldn't have been a good use case either. Code that qualified for an eventual `onComponentMount` should depend on that component being mounted to work. In this case, what we want to run has no connection whatsoever with the component. Similarly, you could run the code from the `propsTransformer` since you know it's going to execute, but that would make it a side-effect again.

File: modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp (hunk 2 above)
Comment (julien) [reply modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp]: Updated in [https://github.com/liferay-frontend/liferay-portal/pull/835/commits/4414fb3b9bfc1ae71b5b6686df5a01126568cefb](https://github.com/liferay-frontend/liferay-portal/pull/835/commits/4414fb3b9bfc1ae71b5b6686df5a01126568cefb)

File: modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/js/EditRoleAssignmentsManagementToolbarPropsTransformer.js (hunk 3 above)
Comment (julien) [reply modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/js/EditRoleAssignmentsManagementToolbarPropsTransformer.js]: Updated in [https://github.com/liferay-frontend/liferay-portal/pull/835/commits/f61edbe6b2760b70463c0cdeee58360999222eff](https://github.com/liferay-frontend/liferay-portal/pull/835/commits/f61edbe6b2760b70463c0cdeee58360999222eff)

File: modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp (hunk 4 above)
Comment (julien) [reply modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp:155]: Yeah it's true that's not really related to the component. 

I'll try with 

```js
<aui:script require='<%= npmResolvedPackageName +  "js/EditRoleAssignments" %>'>
    jsEditRoleAssignments.default({
        ...
    });
</aui:script>
```

Comment (julien) [reply modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp]: The change in [https://github.com/liferay-frontend/liferay-portal/pull/835/commits/ff87e533e5ffe658bdf531af5cf8d5e68fb6cf2f](https://github.com/liferay-frontend/liferay-portal/pull/835/commits/ff87e533e5ffe658bdf531af5cf8d5e68fb6cf2f) works as expected but I'm not sure "this is the way".

I'll squash those extra commits before forwarding, but I thought it would easier to leave them on top during the review.

File: modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp [hunk 5]
```diff
@@ -149,7 +149,14 @@ HashMap<String, Object> additionalProps = HashMapBuilder.<String, Object>put(
...
+<aui:script require='<%= npmResolvedPackageName + "/js/EditRoleAssignments as EditRoleAssignments" %>'>
+	EditRoleAssignments.default({
+		editRoleAssignmentsURL:
+			'<%= additionalProps.get("editRoleAssignmentsURL") %>',
+		modalSegmentState: '<%= additionalProps.get("modalSegmentState") %>',
+		namespace: '<portlet:namespace />',
+		portletURL: '<%= additionalProps.get("portletURL") %>',
+		roleName: '<%= additionalProps.get("roleName") %>',
+		selectAssigneesURL: '<%= additionalProps.get("selectAssigneesURL") %>',
+	});
```
Comment (drewbrokke) [inline modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp:161]: IMO if we are using an inline script, then there is no need for the EditRoleAssignments component to exist. We can just inline the logic and require the `add_assignees` script directly.

Same with this usage of `additionalProps`, we instead just can just pass in the values directly as defined on the JSP.

Comment (julien) [reply modules/apps/roles/roles-admin-web/src/main/resources/META-INF/resources/edit_role_assignments.jsp]: Ok, yes that also makes sense.
I updated that in [https://github.com/liferay-frontend/liferay-portal/pull/835/commits/42f0e6ff589a4545beec59654d1276efa99730c6](https://github.com/liferay-frontend/liferay-portal/pull/835/commits/42f0e6ff589a4545beec59654d1276efa99730c6), let me know what you think when you can.

**Review by kresimir-coko — APPROVED** [review]
Chunky PR, LGTM looking at a glance

**Comment by jbalsas** [comment]
Hey @julien, see you marked this as ready for review, but the "[Don't forward]" in the title might be a bit misleading 😉

**Comment by julien** [comment]
@jbalsas yes sorry, I forgot updating that - thanks for the heads up

**Comment by julien** [comment]
While I'm at it, I'm might squash those "extra" commits and rebase with master.

@drewbrokke since you kindly participated in the review, do you think we could forward this one?

**Comment by julien** [comment]
ci:test:sf

**Comment by julien** [comment]
ci:test:relevant

**Comment by julien** [comment]
> Unable to uninstall com.liferay.account.admin.web_2.0.0 [1276]
java.lang.IllegalStateException: Module has been uninstalled.

Doesn't seem related to these changes. (cc @john-co)

**Comment by calvinctvo** [comment]
ci:test:bundle

**Comment by calvinctvo** [comment]
ci:stable

**Comment by calvinctvo** [comment]
ci:test:stable

**Comment by calvinctvo** [comment]
Hi @julien , the failure doesn't seem to be related to the PR. I've retrigger `ci:test:stable` and it hasn't finish running.

**Comment by julien** [comment]
@calvinctvo thanks for the help :+1: 

@drewbrokke if you have time could you please let me know if we can forward this one? 
Thanks

**Comment by jbalsas** [comment]
ci:forward
